csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -2,195 +2,142 @@
+        private readonly HashSet<IServiceObserverPrx> _observers = new();
+
+        private readonly List<ServiceInfo> _services = new();
         private Communicator? _sharedCommunicator;
+
         private readonly int _traceServiceObserver;
 
         internal ServiceManager(Communicator communicator, string[] args)
         {
             _communicator = communicator;
             _logger = _communicator.Logger;
 
-            _adminEnabled = _communicator.GetPropertyAsBool("Ice.Admin.Enabled") ??
+            bool adminEnabled = _communicator.GetPropertyAsBool("Ice.Admin.Enabled") ??
                 _communicator.GetProperty("Ice.Admin.Endpoints") != null;
 
-            if (_adminEnabled)
+            if (adminEnabled)
             {
-                _adminFacetFilter = new HashSet<string>(_communicator.GetPropertyAsList("Ice.Admin.Facets") ??
-                                                        Array.Empty<string>());
+                _adminFacetFilter = new HashSet<string>(
+                    _communicator.GetPropertyAsList("Ice.Admin.Facets") ?? Array.Empty<string>());
             }
 
             _argv = args;
             _traceServiceObserver = _communicator.GetPropertyAsInt("IceBox.Trace.ServiceObserver") ?? 0;
         }
 
-        public void AddObserver(IServiceObserverPrx? observer, Current current, CancellationToken cancel)
+        public ValueTask AddObserverAsync(IServiceObserverPrx observer, Current current, CancellationToken cancel)
         {
-            // Null observers and duplicate registrations are ignored
-            if (observer != null)
+            lock (_mutex)
             {
-                string[] activeServices;
-                lock (_mutex)
+                if (!_observers.Add(observer))
                 {
-                    try
-                    {
-                        _observers.Add(observer);
-                    }
-                    catch (ArgumentException)
-                    {
-                        return;
-                    }
+                    return default; // ignore duplicates.
+                }
+            }
 
-                    if (_traceServiceObserver >= 1)
-                    {
-                        _logger.Trace("IceBox.ServiceObserver", $"Added service observer {observer}");
-                    }
+            if (_traceServiceObserver >= 1)
+            {
+                _logger.Trace("IceBox.ServiceObserver", $"added service observer {observer}");
+            }
 
-                    activeServices = _services.Where(info => info.Status == ServiceStatus.Started).Select(
-                        info => info.Name).ToArray();
-                }
+            IEnumerable<string> activeServices = _services.Where(info => info.IsActive).Select(info => info.Name);
+
+            if (activeServices.Any())
+            {
+                _ = StartedAsync(observer, activeServices);
+            }
```

Comment:
What prevents a service from being activated or deactivated while this is ran? If a service is deactivated after the `if (activeServices.Any())` but before the `StartedAsync` call the observer might end up getting an incorrect information?

Reply:
That's right. But the existing code suffers from this issue as well.

Reply:
It's a lot less likely to occur with the existing code... and if the observer servant is registered with an OA that enables serialization this problem can only occur in case of a connection failure (which in applications like the IceGrid GUI are considered fatal since they end up tearing down the session).

Now, if the observer is attached while a service is being started/stopped running into this race condition won't be that unlikely... It would be better to provide a reliable way to receive observer updates in order.

Reply:
You're right. Fixed - please re-review.